document/api/src/main/java/io/camunda/document/api/DocumentOperationResponse.java

```diff
@@ -0,0 +1,45 @@
+/*
+ * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH under
+ * one or more contributor license agreements. See the NOTICE file distributed
+ * with this work for additional information regarding copyright ownership.
+ * Licensed under the Camunda License 1.0. You may not use this file
+ * except in compliance with the Camunda License 1.0.
+ */
+package io.camunda.document.api;
+
+public sealed interface DocumentOperationResponse<T> {
+
+  record Success<T>(T result) implements DocumentOperationResponse<T> {}
+
+  record Failure<T>(String message, DocumentErrorCode errorCode, Throwable cause)
+      implements DocumentOperationResponse<T> {
+
+    public Failure(final String message, final DocumentErrorCode errorCode) {
+      this(message, errorCode, null);
+    }
+
+    public Failure(final DocumentErrorCode errorCode, final Throwable cause) {
+      this(null, errorCode, cause);
+    }
+
+    public Failure(final DocumentErrorCode errorCode) {
+      this(null, errorCode, null);
+    }
+
+    public Failure(final String message) {
+      this(message, null, null);
+    }
+
+    public Failure(final Throwable cause) {
+      this(null, null, cause);
+    }
+  }
+
+  enum DocumentErrorCode {
+    DOCUMENT_NOT_FOUND,
+    DOCUMENT_ALREADY_EXISTS,
+    INVALID_INPUT,
+    OPERATION_NOT_SUPPORTED,
+    UNKNOWN_ERROR
+  }
```

Comment:
We could go with a sealed class hierarchy here instead of an enum. This would allow individual errors to return different properties
DOCUMENT_NOT_FOUND: The id of the requested document.
DOCUMENT_NOT_FOUND: The operation that is not supported.
DOCUMENT_ALREADY_EXISTS: Potentially the document that exists already.
INVALID_INPUT: Details about the invalid input. For example validation errors.